text/utilities/0000-health-checker-buildpack.md

```diff
@@ -0,0 +1,47 @@
+# Health Checker Support
+
+## Summary
+
+Introduce a Health Checker buildpack. This buildpack will provide access to one or more health checkers, where a health checker is a small binary that will be installed into the container image for the purpose of validating the health of the runnning image.
+
+By default, we'll provide a tiny HTTP health checker which can be used to probe a port and path. An HTTP 200 OK response returns exit code 0 and a non-2xx response returns 1.
+
+The Health Checker buildpack would be maintained by the existing [Utilities team](https://github.com/orgs/paketo-buildpacks/teams/utilities). It uses libpak & libcnb.
+
+## Motivation
+
+When running buildpack generated images on Kubernetes, HTTP-based health checking is provided by the platform. When running your buildpack generated images with Docker or Docker Compose, there is no option for HTTP-based health checks. It requires a binary in the image to perform the HTTP operation. This is typically `curl` or `wget`, however, the Paketo tiny and base images do not include these binaries.
+
+Having a buildpack that installs a health checker allows for HTTP-based health checks when running on Docker. In addition, it could allow for other types of health checkers for example if your application does not use HTTP.
+
+## Detailed Explanation
+
+1. A health checker buildpack has been developed. It will be contributed to Paketo.
+   1. By default, no health checker will be installed by the buildpack.
+   2. If a user sets `BP_HEALTH_CHECKER_ENABLED=true` then the buildpack will detect and contribute a health check binary to the `$PATH`.
+   3. A process type of `health-check` will be contributed with the command to run the health checker. This may be invoked by the user rather than needing the full command.
+
+## Rationale and Alternatives
+
+1. A user can create a custom stack and include a tool like `curl` or `wget`. This is a lot of additional work, plus `curl` and `wget` are overkill for a simple HTTP health check. To reduce footprint, they are not recommended.
+
+## Implementation
+
+The new Health Checker buildpack will participate if all the following conditions are met
+
+* A user sets `BP_HEALTH_CHECKER_ENABLED=true`
+
+When conditions are met, the buildpack will do the following:
+
+* Contributes a health checker to a layer marked `launch` with command on `$PATH`
+* Contributes a process type called `health-check` which will execute the installed health checker.
```

Comment:
I am assuming we will want runtime configuration options for setting the health url (or host/port/path separately)? If so
1. What should the defaults be?
2. Should other buildpacks in the composite buildpack be able to configure these values (.e.g the sping boot buidpacks sets the URL to the actuator endpoint)?

Reply:
> 1.What should the defaults be?

At the moment, my plan is to have the following config options:

- `BP_HEALTH_CHECKER_ENABLED` -> default false, true means we contribute a health checker
- `BP_HEALTH_CHECKER_DEPENDENCY` -> defaults to `thc` (tiny-health-checker, which is a small http-based health checker I wrote), but can be any dependency id from buildpack.toml's metadata. The idea is that there can be more than one health checker since not every app might provide an HTTP endpoint (could be some other protocol).
- Generically, `BP_HEALTH_CHECKER_<dep-id>_ARGS` would be options you can pass to a given health checker. I have this namespace by the dep id because I think you would want different defaults per health checker. I could possibly see a case for just having `BP_HEALTH_CHECKER_ARGS` that apply to any health checker though, so interested in thoughts on that.

>2. Should other buildpacks in the composite buildpack be able to configure these values (.e.g the sping boot buidpacks sets the URL to the actuator endpoint)?

Right now, I have it set so that an upstream buildpack can request `hc` in the build plan and it'll activate and add a health checker. I did not expose the ability to pick a particular type of health checker or pass args, but I think a buildpack could set that through env variables too, right? If there's a need to configure those through the buildplan, I could add that too.

Do you think this needs to be documented in the RFC? I was thinking the specific options were not super important since this is only intended to be a single implementation, not a standard that multiple buildpacks would implement. I can add it if you want, just let me know.

Reply:
> I did not expose the ability to pick a particular type of health checker or pass args, but I think a buildpack could set that through env variables too, right? If there's a need to configure those through the buildplan, I could add that too.

Conventionally we don't have buildpacks to set `BP_*` environment variables, reserving those for users, instead preferring a `BPI_* (buildpack internal) for buildpack-to-buildpack communication. Although, I don't actually see a practical problem so long as the buildpack uses `.default` to ensure the user-provided options take precedence.

Is the health-checker intended to go at the end or the beginning of the order? If it is at the end environment variables are the correct option for passing config forward. If it is at the beginning we need a build plan mechanism for passing config upstream (relative to execution order). My instincts say having it at the end is better but I don't have a strong argument for why and I could be persuaded to go the other way.

>  Do you think this needs to be documented in the RFC? I was thinking the specific options were not super important since this is only intended to be a single implementation, not a standard that multiple buildpacks would implement.

I would like to see details for the buildpack interface and the `thc` interface included in the RFC. If other buildpacks want to include an http healthcheck wouldn't they be using this interface? If I am understanding correclty other implementations would require additions to this interface?

> Generically, BP_HEALTH_CHECKER_<dep-id>_ARGS would be options you can pass to a given health checker. I have this namespace by the dep id because I think you would want different defaults per health checker. I could possibly see a case for just having BP_HEALTH_CHECKER_ARGS that apply to any health checker though, so interested in thoughts on that.

That makes sense given multiple health check implementations. Can you talk through the tradeoff of multiple implementations as the extension point vs. adding configuration options for a single implementation (e.g. http vs. tcp). This doesn't seem to provide a mechanism for injecting an arbitrary implementation and if we are controlling it a single implementation might be easier? Curious to hear your thoughts.

Reply:
>Is the health-checker intended to go at the end or the beginning of the order?

End. Adding to the RFC.

>My instincts say having it at the end is better but I don't have a strong argument for why and I could be persuaded to go the other way.

Same. I'm not totally sure why, but I tend to add things to the end unless there's a specific reason not to.

>BPI_* (buildpack internal) for buildpack-to-buildpack communication. Although, I don't actually see a practical problem so long as the buildpack uses .default`

I've typically used BPI for moving internal settings from build to runtime (like exec.d proc needs XYZ), things a user shouldn't see/care about. Using `.default` would make sense since the user should always have the final say.

I hadn't really considered the case of another buildpack adding the health check when I was writing this because it requires knowledge about the runtime deployment of the image being generated to really know if you should configure a health check. If you're targeting Docker or something without a health check then add it, otherwise it's noise/wasted space. Only the user really knows that. 

I do think there could be a benefit though, like you said, in having Spring Boot CNB look to see if a health check has been requested (if the env is set to true) and then configure the health checker arguments to look at the actuator endpoint, although that location can be reconfigured so I'd have to look into how we could reliably get that info.

Anyway, the intent is to have it at the end of the order groups so env variables make sense. I'll add a note about buildpack-to-buildpack comms in the RFC and recommend using `.default`.

>I would like to see details for the buildpack interface and the thc interface included in the RFC. If other buildpacks want to include an http healthcheck wouldn't they be using this interface? 

Done.

>If I am understanding correclty other implementations would require additions to this interface?

It depends on how you look at it :) I'm hoping it is sufficient to document the interface as having one `$BP_HEALTH_CHECKER_<dep-id>_ARGS` per health checker implementation that is supported. So the interface is defined as dependent on the name of the health checker dependency.

I guess if we wanted to be more explicit about it, I could put the specific dep id that we plan to support out of the box and we could submit adjustment RFCs to add new health checker implementations. If you feel like we should go that route, it's fine by me.

>Can you talk through the tradeoff of multiple implementations as the extension point vs. adding configuration options for a single implementation (e.g. http vs. tcp). This doesn't seem to provide a mechanism for injecting an arbitrary implementation and if we are controlling it a single implementation might be easier?

My thought for having multiple separate implementations is that it makes the health checkers smaller and simpler. The present one is a static binary just a little under 1M. I haven't done it yet, but a quick test says`upx` would get that down to about 350k. Hopefully there is a day where we can safely ship upx compressed binaries in buildpacks again :) 

Anyway, people seem to really value keeping the size of dependencies small so that was the primary factor in going this route. If we look at adding additional features like Rsocket or gRPC, that would pull in dependencies not everyone needs and making the binary larger.

Aside from that, I was hoping that over time there might be some OSS "standard" or community agreed-upon health checker that pops up and we could use that instead, so we don't have to maintain an actual health checker.